### ISSUE TITLE
Understanding Tensorboard Graph

### ISSUE THREAD

**Issue by qo4on** [tensorflow/tensorboard#3210]
I'm trying to understand graphs in [this tutorial][1]. What are the meaning of nodes it creates? For example, let's take a simple flatten operation `(batch_size, 28, 28) -> (batch_size, 784)`:
[![enter image description here][2]][2]

 - What is the `Const` node and how should I interpret `"tensor_shape":{"dim":[{"size":2}]}` and `"tensor_content":"\\377\\377\\377\\377\\020\\003\\000\\000"}`?

 - What are the `T` and the `Tshape` of the `Reshape` node?
[![enter image description here][3]][3]

 - What is the purpose of the `Identity` node?

Also I understand how `sparse_categorical_crossentropy` and `accuracy` are calculated, but when I expand `metrics` and `loss` it looks horrible. I'm asking mostly not about particular math nodes but about the nodes, which destination is unclear from the description, such as: `keras_learning_phase/input, keras_learning_phase, resource, ReadVariableOp, div_no_nan, dimension, Cast (Truncate {"b":false}), stack...` All the information I found on this topic gives just an overview of the main nodes, but not guts of the graph.

[![enter image description here][5]][5]

  [1]: https://colab.research.google.com/github/tensorflow/tensorboard/blob/master/docs/graphs.ipynb
  [2]: https://i.stack.imgur.com/c8Wrr.png
  [3]: https://i.stack.imgur.com/2FHRQ.png
  [4]: https://i.stack.imgur.com/SAqjU.png
  [5]: https://i.stack.imgur.com/OqAhi.png

**Comment by gowthamkpr** [comment]
@qo4on integers are stored in octal format in graph definition. Thus converting `tensor content = \\377\\377\\377\\377\\020\\003\\000\\000` back to numpy array using `numpy.fromstring(b"\377\377\377\377\020\003\000\000",dtype=np.int32)` yields a numpy array of shape `array([ -1, 784])` and `dtype=int32` 

So const node holds the shape to be applied on input to flatten it.

The attributes of the `flatten/Reshape` T and Tshape are the inputs to the Reshape node i.e., T is the input tensor of shape (?,28,28) which is of type float and Tshape is the shape to be applied on the input tensor.

As far as identity node is considered it is useful when you want to explicitly transport tensor between devices (like, from GPU to a CPU). The node adds send/recv nodes to the graph, which make a copy when the devices of the input and the output are different. (Just remember it makes a copy of the output flattened tensor).

**Comment by qo4on** [comment]
@gowthamkpr Thank you.
How can I know which device the tensor belongs to before and after the identity node? I didn't create it explicitly, `identity` node was made by default.

As far as I can see `resource - ReadVariableOp` pair reads the weights. Why do we need both of them independent?
![image](https://user-images.githubusercontent.com/35609308/73820561-dbf3fc00-4813-11ea-93ec-352e52f0670a.png)

**Comment by gowthamkpr** [comment]
I am not sure about the devices part but,

> why there are 3 inputs in the identity node after relu(input * weights + bias)?
As you can see the identity node has two control dependencies `dense/bias/ReadVariableOp` and `dense/bias/ReadVariableOp` which means these two inputs are not passing data into the identity node.
(Control dependency definition: A statement S2 is control dependent on S1 (written ) if and only if S2's execution is conditionally guarded by S1). So, effective there is only one data input to identity node, i.e., dense/Relu

> What is the purpose of pair Input - keras_learning_phase?
The learning phase flag is a bool tensor (0 = test, 1 = train) to be passed as input to any Keras function that uses a different behavior at train time and test time i.e., droupout can only occur during training phase.

**Comment by qo4on** [comment]
Thanks.
Let's see if anybody knows the `identity` node meaning.

> How to find out which device the tensor belongs to before and after the `identity` node?

Understanding this is very useful.

**Comment by wchargin** [comment]
> > How to find out which device the tensor belongs to before and after
> > the `identity` node?
> 
> Understanding this is very useful.

In the left panel of the graph dashboard, you should be able to set the
“Color” selector to “Device” to color-code nodes by the device on which
they execute. A legend will appear right under the radio buttons.

**Comment by gowthamkpr** [comment]
Closing this issue as it has been inactive for a long time. Please add additional comments for us to open this issue again. Thanks!
